### PR TITLE
[Market+GPoS+Balance] Add transferrable_balance into balance module

### LINKED ISSUE #164: [GPoS+Market] Pledge and bond should be mutex

### PR DESCRIPTION
**Pull request by MyronFanQiu**
Fix #164

### CODE REVIEW

File: cstrml/balances/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1313 @@
...
+	}
+}
+
+impl<T: Trait<I>, I: Instance> TransferrableCurrency<T::AccountId> for Module<T, I>
+where
+	T::Balance: MaybeSerializeDeserialize + Debug
+{
+	fn transferrable_balance(who: &T::AccountId) -> Self::Balance
+	{
+		let sum = Self::locks(who).iter().fold(Zero::zero(), |acc: Self::Balance, value| {acc + value.amount});
```
Comment (zikunfan) [inline cstrml/balances/src/lib.rs:1301]: `sum` to `locks`?

File: cstrml/staking/src/lib.rs [hunk 2]
```diff
@@ -727,26 +730,23 @@ decl_module! {
...
-                // Candidates should judge its stake limit, this promise candidates' bonded stake
-                // won't exceed.
-                if <Validators<T>>::contains_key(&stash) {
-                    let limit = Self::stake_limit(&stash).unwrap_or_default();
-                    if ledger.active >= limit {
-                        Err(Error::<T>::NoWorkloads)?
-                    } else {
-                        extra = extra.min(limit - ledger.active);
-                    }
+            let mut extra = T::Currency::transferrable_balance(&stash);
```
Comment (zikunfan) [inline cstrml/staking/src/lib.rs:733]: Do we need add more `tests` to cover the *double lock* scenario?

File: cstrml/balances/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1313 @@
...
+		id: LockIdentifier,
+		who: &T::AccountId,
+	) {
+		let mut locks = Self::locks(who);
+		locks.retain(|l| l.id != id);
+		Self::update_locks(who, &locks[..]);
+	}
+}
+
+impl<T: Trait<I>, I: Instance> TransferrableCurrency<T::AccountId> for Module<T, I>
```
Comment (zikunfan) [inline cstrml/balances/src/lib.rs:1295]: Do we need to add some tests inside `balance` pallet?

File: cstrml/staking/src/lib.rs (hunk 2 above)
Comment (MyronFanQiu) [reply cstrml/staking/src/lib.rs:733]: I would add some tests into balance module

File: cstrml/balances/src/lib.rs (hunk 1 above)
Comment (MyronFanQiu) [reply cstrml/balances/src/lib.rs:1301]: rename it to `sum_of_locks`

File: cstrml/balances/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+					let _ = Balances::slash(&1, 1);
+					// The account should be dead.
+					assert!(Balances::is_dead_account(&1));
+					assert_eq!(Balances::free_balance(1), 0);
+					assert_eq!(Balances::reserved_balance(1), 0);
+				});
+		}
+
+		#[test]
+		fn account_transfer_balance_works() {
```
Comment (MyronFanQiu) [inline cstrml/balances/src/tests.rs:662]: This is the added test

Comment (zikunfan) [reply cstrml/balances/src/tests.rs:662]: Rename to `account_transfer_balance_should_work`?

**Pull request merged**